lib/analysisPoller.js

```diff
@@ -57,6 +57,9 @@ async function handleErrors (response, accessToken, uuid) {
   if (status < 200 || status > 299) {
     let msg = `Failed in retrieving analysis response, HTTP status code: ${status}. UUID: ${uuid}`
     switch (status) {
+      case 401:
+        msg = response
```

Comment:
@fgimenez This is wrong. `response` at this point is this object: https://developer.mozilla.org/en-US/docs/Web/API/Response
To get the real human-readable message reported by API, we need to:
```
msg = (await res.json()).error;
```

Reply:
Actually, this entire function can be replaced by such one-liner, if we don't care about keeping current error messages intract in Armlet.

Reply:
Right now the results I have seen from the API side are disappointing,  to say the least. So for now let's keep them and revisit _after_ the API is done. Thanks.

Reply:
@rocky The main point of my comment here is that doing `msg = response` you assign JS Object to a variable that is expected to be the error description String. And that JS Object is the HTTP response: even if it gets auto stringified further down the line, resulting string will be very different from a human-readable msg you expect.

Reply:
If we have to nuke the error thing further down to make everything a string great. 

Get used to the idea that in those situations where we don't want tracebacks we don't create them and then rely on throwing them away.

Reply:
@birdofpreyru @rocky the code that consumes this module expected it to throw an object when a 401 happen, it checked then the statusCode and refreshed the tokens if needed. When this code was changed to throw a string instead of an object the consumer code failed to refresh the tokens properly.

I agree that throwing an object in some cases and a string in other cases is not ok, this was a quick fix to unblock beta testing. I like the proposal of @birdofpreyru of handling refresh inside the different client modules, let me know if you are up to work on the refactor.

WRT throwing a string instead of an error object I still don't understand why is this needed. The consumer code doesn't need to blindly print whatever is thrown, if it is an error object the human-readable info can be in the `message` property, no need to show any traceback. Moreover, throwing an object instead of a string literal allow the consumer code to type check the value returned, instead of doing some esoteric parsing on string to know what happened. So i don't really understand why is it better to throw a string instead of a error object, am i missing something?

Reply:
@fgimenez @birdofpreyru thanks for the clarifications. 

Right now I need to focus on how to integrate better with truffle and its artifiacts and this discussion is taking away from that time. 

As is the way I work, if others feel strongly to work on something, that that is good, I am grateful,  and by all means go ahead!  I'm a high-level guy here. 

The main thing I ask is:

* do not create useless tracebacks when they are not wanted
*  remove useless tracebacks as early as possible
* make it so that users of this library aren't confused by there being a traceback field even if empty
* keep the ability, even if in stub form, to be able to override and rewrite responses from the API. (currrently we are in some places filling in information that aren't available in the reponses)

I hope these overall guidelines make sense. If not, let me know. As along as all of these are met, the feel free to redo in the best way you see fit. Between the two (or three including Daniyar), I will be honest and say that you know more about this and the best way to do this than I. 

So whatever the two or three of you come up with you are, as they say in the credit-card business "pre-approved" and I don't want to get in the way (and probalby don't have time to either). 

So yes, if among you, you see how to do a refactor properly to fix up the mess I made, by all means please do.  Thanks in advance.

Reply:
> WRT throwing a string instead of an error object I still don't understand why is this needed.

It is definitely better to always throw Error objects, i.e. `throw new Error(message)`; but existing Armlet code expects string to be thrown. And yeah, what you are saying that for some errors it expects objects to be thrown - that's definitely bad it follows different conventions for different error codes.

> I like the proposal of @birdofpreyru of handling refresh inside the different client modules, let me know if you are up to work on the refactor.

Is it question to me? I can do such refactoring, but most probably not this week.

Reply:
> remove useless tracebacks as early as possible

ehh... I would not call tracebacks useless: whenever anything goes wrong a detailed traceback of error is your best friend :)

Reply:
`throw new Error(message)` creates a traceback, right? If so, that's not following the guideline.

It was a general question. As long as things refresh and/or functionality isn't broken, then from my standpoint this can wait indefinitely.

Reply:
Tracebacks are good! If in any place you don't want to show traceback to end user, just catch the error and print `error.message` instead of `error` itself (`.toString()` method of `Error` class, triggered by attempt to just `console.log(error)` includes traceback into the result).

Reply:
We've discussed this before. Not here. This is not up for discussion.

Reply:
I missed that discussion. But it is, indeed, not a thing to discuss. Throwing Error objects is considered a good practice in JS community: https://eslint.org/docs/rules/no-throw-literal

Reply:
Also:
https://stackoverflow.com/questions/11502052/throwing-strings-instead-of-errors

Reply:
It is especially true for SDK as Armlet, as thrown errors may propagate to the host code, and the host code in 99% of cases will expect to get Error object, not a string, and definitely not different stuff for different errors.

Reply:
@birdofpreyru i tried to answer a previous question from you but looks like this conversation has been censored! :laughing: 
>> I like the proposal of @birdofpreyru of handling refresh inside the different client modules, let me know if you are up to work on the refactor.

> Is it question to me? I can do such refactoring, but most probably not this week.

ok, i'll try to work on it on monday, will let you know how it goes.

Reply:
There has been _no_ censorship that I am aware of. 

I ask you though to stay within the guidelines metioned above which includes: no tracebacks, at least coming from lib/analysisPoller.js. 

The checked answer in one of the SO  links says:
> It is okay to throw whatever you like, but keep in mind that if the catch is outside of your own code, it might expect a full Error instance and not a plain string.

"might" is different than "must".   When either of you writes a client that uses armlet and can show me how it makes superor use of Error instances, then that will be interesting,. Right now I've written two and Error instances is not helpful.

By all means throw an Error instance when an Error instance maes sense. And when it doesn't, don't. 

Think of that analogous to physical laws where rules get refined over time depending the specific situtation. Newtonian laws of motion works fine as a general approximation for most situtations, but in certain situations motion taking into account General Relativity is better. 

So in the context of lib/analysisPoller.js throwing strings without tracebacks has observed to be more useful. Please respect that. 

Also, as I have said discussing this when I am trying to finish something is is becoming counterproductive. Please keep in mind that this is not just about whether you have tihe time to dicuss it but also whether I do as well.

Reply:
@rocky not sure why you keep talking about throwing strings instead of error objects, my comment above was about refactoring the refresh code in the way Sergey mentioned. 

Regarding censorship, this morning i tried to answer a question from Sergey that i had in email but the message was not here. Thankfully the full thread is here again, maybe that's why you keep explaining why you think `console.log(e)` is much less work than `console.log(e.message)`?

I won't spend time discussing this given that you are trying to finish something, ping me when ready.

Reply:
> not sure why you keep talking about throwing strings instead of error objects, my comment above was about refactoring the refresh code in the way Sergey mentioned.

Ok - my misunderstanding. I apologize. I am and have always been been for refactoring the refresh code in the way Sergey mentioned since he mentioned it. That part is great!

I will ping you when I'm done with what I am currently working on which I don't think will be too long. 

Thanks for your understanding and patience.  And I  _do_ appreciate all the help from both of you too.